bpf/lib/encrypt.h

```diff
@@ -205,6 +206,131 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 #endif /* ENABLE_ROUTING */
 }
 
+/* checks whether a IPsec redirect should be performed for the security id
+ * we do not IPsec encrypt:
+ * 1. Host-to-Host or Pod-to-Host traffic
+ * 2. Traffic leaving the cluster
+ * 3. Remote nodes including Kube API server
+ * 4. Traffic is already ESP encrypted
+ */
+static __always_inline int
+ipsec_redirect_sec_id_ok(__u32 sec_id, int ip_proto) {
+	if (ip_proto == IPPROTO_ESP)
+		return 0;
+	if (sec_id == HOST_ID)
+		return 0;
+	if (!identity_is_cluster(sec_id))
+		return 0;
+	if (identity_is_remote_node(sec_id))
+		return 0;
+	return 1;
+}
+
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
+{
+	struct remote_endpoint_info __maybe_unused *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	struct ipv6hdr __maybe_unused *ip6;
+	__u32 magic __maybe_unused = 0;
+	int ip_proto = 0;
+	int ret = 0;
+	union macaddr dst_mac = CILIUM_NET_MAC;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+		ip_proto = ip4->protocol;
+
+		dst = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		src = lookup_ip4_remote_endpoint(ip4->saddr, 0);
+
+#  if defined(TUNNEL_MODE)
+		/* tunnel mode needs a bit of special handling when
+		 * encapsulated packets get here the destination address is
+		 * already a cluster node IP.
+		 *
+		 * the security ID is appended to the mark in the overlay prog
+		 * and we can extract this with 'get_identity'.
+		 * additionally, this is a VXLAN packet so ip4->daddr is the ip
+		 * of the destination host already and can be passed into
+		 * set_ipsec_encrypt to obtain the correct node ID and spi.
+		 */
+		if (ctx_is_overlay(ctx)) {
+			/* if this is already encrypted, just pass it to stack */
+			if (ctx_is_overlay_encrypted(ctx))
```

Comment:
I don't think this is robust. On upgrade from v1.16 (or an older v1.17), there inherently will be ESP-in-VXLAN packets in flight that now pass through this logic, but don't carry the new `MARK_MAGIC_OVERLAY_ENCRYPTED` yet (because they were handled by an old `to-overlay` program).

Reply:
This is a good case, but since any in-flight stale packets which are not marked would be ESP-in-VXLAN, and ESP is on the outside:


```diff
			/* if this is already encrypted, just pass it to stack
			 * we check ip_proto for ESP for the v1.16 -> v1.17
			 * upgrade case where stale EinV overlay packets may be
			 * inflight, prior to the v1.17 to-overlay code which
			 * marks encrypted packets being tunnel encap'd.
			 */
			if (ctx_is_overlay_encrypted(ctx) || ip_proto == IPPROTO_ESP)
				return CTX_ACT_OK;
```

Would simply fix this concern, right?

Reply:
yep, and I needed to double-check initially that this check doesn't exist :smile:. Do we then still need the mark? For upgrade to v1.18 maybe? Or should v1.18 just do the same check?

Reply:
fwiw, I think we could just bail out the moment that we see `IPPROTO_ESP` when we take it from `ip4->protocol` / `ip6->ip6->nexthdr`. And thus cover this case *and* the native routing scenario, that otherwise is handled further down in `ipsec_redirect_sec_id_ok()`.

Reply:
> Do we then still need the mark?

we do, so that to-overlay can say "hey don't encrypt this overlay traffic, v1.16 won't be able to handle it" :+1:

Reply:
> yep, and I needed to double-check initially that this check doesn't exist 😄. Do we then still need the mark? For upgrade to v1.18 maybe? Or should v1.18 just do the same check for IPPROTO_ESP?

I maybe still possible to catch VinE traffic from a v1.18 to v1.17 downgrade in flight, in this case, VXLAN is on the outside but the packet would have been marked from overlay and we would not double encrypt.

Reply:
Mistakes were made in this thread lol.
EinV traffic has VXLAN on the outside, I had it backwards in this chat. 
So scrap this, we'll need to write some parsing code to get this right for v1.16 -> v1.17 scenarios.

Reply:
This is now update to parse the inner IPv4 header and determine if the payload is ESP encrypted. Lets run this thru CI.